docs/design-documents/20240812-recover-from-pipeline-errors.md

```diff
@@ -0,0 +1,194 @@
+  recoverable. A connector developer must explicitly mark an error as fatal if
+  it should cause the pipeline to stop permanently.
+
+## Recovery mechanism
+
+### Backoff delay strategy
+
+To manage transient errors and prevent constant pipeline restarts, Conduit will
+implement a backoff delay strategy. This strategy introduces a waiting period
+before attempting to restart the pipeline, allowing time for issues to resolve
+themselves. The backoff parameters will be configurable to allow customization
+based on specific production needs.
+
+- **Configurable Parameters:**
+  - **Minimum Delay Before Restart:** Default: 1 second
+  - **Maximum Delay Before Restart:** Default: 1 minute
+  - **Backoff Factor:** Default: 2
+  - **Maximum Number of Retries:** Default: Infinite
+  - **Delay After Which Fail Count is Reset:** Default: 5 minutes (or on first
+    successfully end-to-end processed record)
+
+This results in a default delay progression of 1s, 2s, 4s, 8s, 16s, 32s, 1m,...,
+balancing the need for recovery time and minimizing downtime.
+
+### Consecutive failures and permanent stop
+
+To avoid endless restarts, Conduit will include an option to permanently stop
+the pipeline after a configurable number of consecutive failures. Once the
+pipeline successfully processes at least one record or runs without encountering
+an error for a predefined time frame, the count of consecutive failures will
+reset.
+
+## Pipeline state management
+
+A new pipeline state, `recovering`, will be introduced to indicate that a
+pipeline has experienced a fault and is in the process of recovering. This state
+will help differentiate between an actively running pipeline and one that is in
+a restart loop.
+
+- **State Transition:**
+  - `running` → `recovering` → `running` (after pipeline runs without failure
+    for the specified duration or a record is processed successfully end-to-end)
+  - `running` → `recovering` → `degraded` (if max retries reached or fatal error
+    occurs)
+
+This approach will improve the visibility of pipeline status and assist in
+debugging and monitoring.
+
+## Record deduplication
+
+Restarting a failed pipeline can result in some records being re-delivered, as
+some records might have already been processed by the destination, but the
+acknowledgments might not have reached the source before the failure occurred.
+For that case, we should add a mechanism in front of the destination which
+tracks records that have been successfully processed by the destination, but the
+acknowledgment wasn't processed by the source yet. If such a record is
+encountered after restarting the pipeline, Conduit should acknowledge it without
+sending it to the destination.
+
+## Audit log
+
+If a pipeline is automatically recovered, a user needs to check the Conduit logs
+to see the cause and what actually happened. Ideally that information would be
+retrievable through the API and would be attached to the pipeline itself so a
```

Comment:
>  Ideally that information would be retrievable through the API

I guess you mean the Conduit API? How about we add this to the metrics? It might be easier to integrate with (because of existing tools that use those metrics), will provide value and it's also relatively easy to implement.

Reply:
Yes, I meant the Conduit API (HTTP, gRPC). I'm not following though, how would you return that information through metrics? 🤔 I'm talking about having a log of events that happened on a pipeline. Or maybe a global log of events, so that actions like deleting a pipeline are also recorded, even if the pipeline doesn't exist anymore.

Reply:
We discussed this in a meeting, we'll add some metrics in the first iteration.